src/zkevm_specs/evm_circuit/execution/create.py

```diff
@@ -0,0 +1,186 @@
+from zkevm_specs.util.arithmetic import FQ, RLC
+from zkevm_specs.util.hash import EMPTY_CODE_HASH
+from zkevm_specs.util.param import (
+    GAS_COST_COPY_SHA3,
+    GAS_COST_CREATE,
+    N_BYTES_ACCOUNT_ADDRESS,
+    N_BYTES_GAS,
+    N_BYTES_MEMORY_SIZE,
+    N_BYTES_U64,
+)
+from ...util import (
+    CALL_CREATE_DEPTH,
+)
+from ..instruction import Instruction, Transition
+from ..table import RW, CallContextFieldTag, AccountFieldTag, CopyDataTypeTag
+
+
+def create(instruction: Instruction):
+    # check opcode is CREATE or CREATE2
+    opcode = instruction.opcode_lookup(True)
+    is_create, is_create2 = instruction.pair_select(opcode, Opcode.CREATE, Opcode.CREATE2)
+    instruction.responsible_opcode_lookup(opcode)
+
+    # set the caller_id to the current rw_counter
+    callee_call_id = instruction.curr.rw_counter
+
+    # Stack parameters and result
+    value = instruction.stack_pop()
+    offset_rlc = instruction.stack_pop()
+    size_rlc = instruction.stack_pop()
+    salt_rlc = instruction.stack_pop() if is_create2 == FQ(1) else RLC(0)
+    return_contract_address = instruction.stack_push()
+
+    offset = instruction.rlc_to_fq(offset_rlc, N_BYTES_U64)
+    size = instruction.rlc_to_fq(size_rlc, N_BYTES_U64)
+
+    depth = instruction.call_context_lookup(CallContextFieldTag.Depth)
+    tx_id = instruction.call_context_lookup(CallContextFieldTag.TxId)
+    caller_address = instruction.call_context_lookup(CallContextFieldTag.CallerAddress)
+    nonce, nonce_prev = instruction.account_write(caller_address, AccountFieldTag.Nonce)
+    is_success = instruction.call_context_lookup(CallContextFieldTag.IsSuccess)
+    is_static = instruction.call_context_lookup(CallContextFieldTag.IsStatic)
+    reversion_info = instruction.reversion_info()
+
+    # calculate contract address
+    contract_address = (
+        instruction.generate_contract_address(caller_address, nonce)
+        if is_create == 1
+        else instruction.generate_CREAET2_contract_address(
+            caller_address, salt_rlc.le_bytes, instruction.next.code_hash
+        )
+    )
+
+    # verify return contract address
+    instruction.constrain_equal(
+        instruction.rlc_to_fq(return_contract_address, N_BYTES_ACCOUNT_ADDRESS),
+        is_success * contract_address,
+    )
+
+    # Verify depth is less than 1024
+    instruction.range_lookup(depth, CALL_CREATE_DEPTH)
+
+    # ErrNonceUintOverflow constraint
+    (is_not_overflow, _) = instruction.compare(nonce, nonce_prev, N_BYTES_U64)
+    instruction.is_zero(is_not_overflow)
+
+    # add contract address to access list
+    instruction.add_account_to_access_list(tx_id, contract_address)
+
+    # ErrContractAddressCollision constraint
+    # code_hash_prev could be either 0 or EMPTY_CODE_HASH
+    # code_hash should be EMPTY_CODE_HASH to make sure the account is created properly
+    code_hash, code_hash_prev = instruction.account_write(
+        contract_address, AccountFieldTag.CodeHash
+    )
```

Comment:
We should revert this account creation in case current context is not persistent.

```suggestion
    code_hash, code_hash_prev = instruction.account_write(
        contract_address, AccountFieldTag.CodeHash, reversion_info
    )
```

Reply:
fixed in [d675ec9](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/355/commits/d675ec97ed971ec044a5bc5a40db3ab57551c629)